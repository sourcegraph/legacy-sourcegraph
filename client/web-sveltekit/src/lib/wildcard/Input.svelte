<script lang="ts">
    import { onMount } from 'svelte'
    import type { FormEventHandler } from 'svelte/elements'

    export let value: string
    export let placeholder: string | undefined
    export let autofocus: boolean | undefined
    export let onInput: FormEventHandler<HTMLInputElement> | undefined = undefined
    export let input: HTMLInputElement | undefined = undefined
    export let actions: Array<(node: HTMLInputElement) => unknown>

    $: bindAction = function bindAction(node: HTMLInputElement) {
        if (actions.length === 0) {
            return
        }

        const actionsResults = actions.map(action => action(node))

        return {
            destroy() {
                actionsResults.map(result => result?.destroy?.())
            },
        }
    }

    onMount(() => {
        if (autofocus) {
            requestAnimationFrame(() => {
                input?.focus()
            })
        }
    })
</script>

<input
    bind:this={input}
    type="text"
    use:bindAction
    {value}
    {autofocus}
    {placeholder}
    on:input={onInput}
    {...$$restProps}
/>

<style lang="scss">
    input {
        display: block;
        width: 100%;
        height: var(--input-height);
        padding: var(--input-padding-y) var(--input-padding-x);
        color: var(--input-color);
        font-size: var(--input-font-size);
        font-weight: var(--input-font-weight);
        line-height: var(--input-line-height);
        background-color: var(--input-bg);
        background-clip: padding-box;
        border: var(--input-border-width) solid var(--input-border-color);
        border-radius: var(--border-radius);
    }
</style>
