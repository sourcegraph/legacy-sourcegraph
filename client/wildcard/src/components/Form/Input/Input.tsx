import { useRef, forwardRef, InputHTMLAttributes, ReactNode } from 'react'

import classNames from 'classnames'
import { useMergeRefs } from 'use-callback-ref'

import { Typography } from '../..'
import { useAutoFocus } from '../../../hooks/useAutoFocus'
import { ForwardReferenceComponent } from '../../../types'

export enum InputStatus {
    initial = 'initial',
    error = 'error',
    valid = 'valid',
}

export interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
    /** text label of input. */
    label?: ReactNode
    /** Description block shown below the input. */
    message?: ReactNode
    /** Custom class name for root label element. */
    className?: string
    /** Custom class name for input element. */
    inputClassName?: string
    /** Exclusive status */
    status?: InputStatus | `${InputStatus}`
    error?: ReactNode
    /** Disable input behavior */
    disabled?: boolean
    /** Determines the size of the input */
    variant?: 'regular' | 'small'
    /** Supports appending an element to input */
    inputSymbol?: ReactNode
    /** Custom class name for input element and symbol wrapper. */
    inputSymbolWrapperClassName?: string
}

/**
 * Displays the input with description, error message, visual invalid and valid states.
 * Does not support Loader icon and status=loadind (user FormInput to get support for loading state)
 */
export const Input = forwardRef((props, reference) => {
    const {
        as: Component = 'input',
        type = 'text',
        variant = 'regular',
        label,
        message,
        className,
        inputClassName,
        disabled,
        status = InputStatus.initial,
        error,
        autoFocus,
        inputSymbol,
        inputSymbolWrapperClassName,
        ...otherProps
    } = props

    const localReference = useRef<HTMLInputElement>(null)
    const mergedReference = useMergeRefs([localReference, reference])

    useAutoFocus({ autoFocus, reference: localReference })
    const inputElement = (
        <Component
            disabled={disabled}
            type={type}
            className={classNames(inputClassName, 'form-control', 'with-invalid-icon', {
                'is-valid': status === InputStatus.valid,
                'is-invalid': error || status === InputStatus.error,
                'form-control-sm': variant === 'small',
            })}
            {...otherProps}
            ref={mergedReference}
            autoFocus={autoFocus}
        />
    )

    const inputWithSymbol = (
        <div className={classNames('d-flex', inputSymbolWrapperClassName)}>
            {inputElement}
            {inputSymbol}
        </div>
    )

    const messageClassName = 'form-text font-weight-normal mt-2'
    const inputWithMessage = (
        <>
            {inputSymbol ? inputWithSymbol : inputElement}
            {error && (
                <small role="alert" className={classNames('text-danger', messageClassName)}>
                    {error}
                </small>
            )}
            {!error && message && <small className={classNames('text-muted', messageClassName)}>{message}</small>}
        </>
    )

    if (label) {
        return (
            <Typography.Label className={classNames('w-100', className)}>
                {label && <div className="mb-2">{variant === 'regular' ? label : <small>{label}</small>}</div>}
                {inputWithMessage}
            </Typography.Label>
        )
    }

    return inputWithMessage
}) as ForwardReferenceComponent<'input', InputProps>

Input.displayName = 'Input'
