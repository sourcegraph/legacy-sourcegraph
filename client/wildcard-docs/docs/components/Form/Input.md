---
id: "Input"
title: "Input"
sidebar_label: "Input"
sidebar_position: 0
custom_edit_url: null
---


• **Input**: `ForwardReferenceComponent`<``"input"``, `InputProps`\>

## Overview

Displays the input with description, error message, visual invalid and valid states.

#### Defined in

[client/wildcard/src/components/Form/Input/Input.tsx:31](https://github.com/sourcegraph/sourcegraph/blob/49e75f130e/client/wildcard/src/components/Form/Input/Input.tsx#L31)

## Usage
```tsx
import { Input } from '@sourcegraph/wildcard'

<Input
    value={selected}
    label="Input small"
    onChange={handleChange}
    message="random message"
    status="valid"
    disabled={false}
    placeholder="testing this one"
    variant="small"
/>
```

## Props

### label

• `Optional` **label**: `string`

text label of input.

#### Defined in

[client/wildcard/src/components/Form/Input/Input.tsx:11](https://github.com/sourcegraph/sourcegraph/blob/49e75f130e/client/wildcard/src/components/Form/Input/Input.tsx#L11)

---

### message
• `Optional` **message**: `ReactNode`

Description block shown below the input.

#### Defined in

[client/wildcard/src/components/Form/Input/Input.tsx:13](https://github.com/sourcegraph/sourcegraph/blob/49e75f130e/client/wildcard/src/components/Form/Input/Input.tsx#L13)

---

### className
• `Optional` **className**: `string`

Custom class name for input element.

#### Defined in

[client/wildcard/src/components/Form/Input/Input.tsx:15](https://github.com/sourcegraph/sourcegraph/blob/49e75f130e/client/wildcard/src/components/Form/Input/Input.tsx#L15)

---

### inputClassName
• `Optional` **inputClassName**: `string`

Custom class name for input element.

#### Defined in

[client/wildcard/src/components/Form/Input/Input.tsx:17](https://github.com/sourcegraph/sourcegraph/blob/49e75f130e/client/wildcard/src/components/Form/Input/Input.tsx#L17)

---

### inputSymbol
• `Optional` **inputSymbol**: `ReactNode`

Input icon (symbol) which render right after the input element.

#### Defined in

[client/wildcard/src/components/Form/Input/Input.tsx:19](https://github.com/sourcegraph/sourcegraph/blob/49e75f130e/client/wildcard/src/components/Form/Input/Input.tsx#L19)

---

### status
• `Optional` **status**: ``"error"`` \| ``"loading"`` \| ``"valid"``

Exclusive status

#### Defined in

[client/wildcard/src/components/Form/Input/Input.tsx:21](https://github.com/sourcegraph/sourcegraph/blob/49e75f130e/client/wildcard/src/components/Form/Input/Input.tsx#L21)

---

### disabled
• `Optional` **disabled**: `Boolean`

Disable input behavior

#### Defined in

[client/wildcard/src/components/Form/Input/Input.tsx:23](https://github.com/sourcegraph/sourcegraph/blob/49e75f130e/client/wildcard/src/components/Form/Input/Input.tsx#L23)

---

### variant

• `Optional` **variant**: ``"regular"`` \| ``"small"``

 Determines the size of the input

 #### Defined in

[client/wildcard/src/components/Form/Input/Input.tsx:25](https://github.com/sourcegraph/sourcegraph/blob/49e75f130e/client/wildcard/src/components/Form/Input/Input.tsx#L25)
