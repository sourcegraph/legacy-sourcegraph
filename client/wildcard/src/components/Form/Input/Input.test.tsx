import { render } from '@testing-library/react'

import { Input } from './Input'

const INPUT_STATUS = ['error', 'valid'] as const

describe('Input', () => {
    describe('Input - does not support loading state', () => {
        it('renders an input correctly', () => {
            const { container } = render(
                <Input
                    defaultValue="Input value"
                    title="Input"
                    label="Input label"
                    message="random message"
                    status="initial"
                    placeholder="initial status input"
                />
            )

            expect(container.firstChild).toMatchInlineSnapshot(`
                <label
                  class="label w-100"
                >
                  <div
                    class="mb-2"
                  >
                    Input label
                  </div>
                  <input
                    class="form-control with-invalid-icon"
                    placeholder="initial status input"
                    title="Input"
                    type="text"
                    value="Input value"
                  />
                  <small
                    class="text-muted form-text font-weight-normal mt-2"
                  >
                    random message
                  </small>
                </label>
            `)
        })

        it.each(INPUT_STATUS)("Renders Input status '%s' correctly", status => {
            const { container } = render(<Input status={status} defaultValue="" />)
            expect(container.firstChild).toMatchSnapshot()
        })
    })
    describe('FormInput - supports loading state', () => {
        it('renders an input correctly', () => {
            const { container } = render(
                <Input
                    defaultValue="Input value"
                    title="Valid input"
                    message="random message"
                    status="valid"
                    placeholder="loading status input"
                />
            )

            expect(container.firstChild).toMatchInlineSnapshot(`
                <input
                  class="form-control with-invalid-icon is-valid"
                  placeholder="loading status input"
                  title="Valid input"
                  type="text"
                  value="Input value"
                />
            `)
        })

        it('renders an input with label correctly', () => {
            const { container } = render(
                <Input
                    defaultValue="Input value"
                    title="Error input"
                    message="random message"
                    status="error"
                    placeholder="error status input"
                    label="Input label"
                />
            )

            expect(container.firstChild).toMatchInlineSnapshot(`
                <label
                  class="label w-100"
                >
                  <div
                    class="mb-2"
                  >
                    Input label
                  </div>
                  <input
                    class="form-control with-invalid-icon is-invalid"
                    placeholder="error status input"
                    title="Error input"
                    type="text"
                    value="Input value"
                  />
                  <small
                    class="text-muted form-text font-weight-normal mt-2"
                  >
                    random message
                  </small>
                </label>
            `)
        })

        it.each(INPUT_STATUS)("Renders FormInput status '%s' correctly", status => {
            const { container } = render(<Input status={status} defaultValue="" />)
            expect(container.firstChild).toMatchSnapshot()
        })
    })
})
