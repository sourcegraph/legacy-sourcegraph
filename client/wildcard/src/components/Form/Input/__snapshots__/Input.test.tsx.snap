// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input FormInput - supports loading state Renders FormInput status 'error' correctly 1`] = `
<input
  class="form-control with-invalid-icon is-invalid"
  type="text"
  value=""
/>
`;

exports[`Input FormInput - supports loading state Renders FormInput status 'valid' correctly 1`] = `
<input
  class="form-control with-invalid-icon is-valid"
  type="text"
  value=""
/>
`;

exports[`Input Input - does not support loading state Renders Input status 'error' correctly 1`] = `
<input
  class="form-control with-invalid-icon is-invalid"
  type="text"
  value=""
/>
`;

exports[`Input Input - does not support loading state Renders Input status 'valid' correctly 1`] = `
<input
  class="form-control with-invalid-icon is-valid"
  type="text"
  value=""
/>
`;
